tasks.rb

```diff
@@ -4916,6 +4933,210 @@ def download_epws
+
+  base_elements_xsd = File.read(File.join(File.dirname(__FILE__), 'HPXMLtoOpenStudio', 'resources', 'BaseElements.xsd'))
+  base_elements_xsd_doc = Oga.parse_xml(base_elements_xsd)
+
+  # construct dictionary for enumerations and min/max values of HPXML data types
+  hpxml_data_types_xsd = File.read(File.join(File.dirname(__FILE__), 'HPXMLtoOpenStudio', 'resources', 'HPXMLDataTypes.xsd'))
+  hpxml_data_types_xsd_doc = Oga.parse_xml(hpxml_data_types_xsd)
+  hpxml_data_types_dict = {}
+  hpxml_data_types_xsd_doc.xpath('//xs:simpleType | //xs:complexType').each do |simple_type_element|
+    enums = []
+    simple_type_element.xpath('xs:restriction/xs:enumeration').each do |enum|
+      enums << enum.get('value')
+    end
+    minInclusive_element = simple_type_element.at_xpath('xs:restriction/xs:minInclusive')
+    min_inclusive = minInclusive_element.get('value') if not minInclusive_element.nil?
+    maxInclusive_element = simple_type_element.at_xpath('xs:restriction/xs:maxInclusive')
+    max_inclusive = maxInclusive_element.get('value') if not maxInclusive_element.nil?
+    minExclusive_element = simple_type_element.at_xpath('xs:restriction/xs:minExclusive')
+    min_exclusive = minExclusive_element.get('value') if not minExclusive_element.nil?
+    maxExclusive_element = simple_type_element.at_xpath('xs:restriction/xs:maxExclusive')
+    max_exclusive = maxExclusive_element.get('value') if not maxExclusive_element.nil?
+
+    simple_type_element_name = simple_type_element.get('name')
+    hpxml_data_types_dict[simple_type_element_name] = {}
+    hpxml_data_types_dict[simple_type_element_name][:enums] = enums
+    hpxml_data_types_dict[simple_type_element_name][:min_inclusive] = min_inclusive
+    hpxml_data_types_dict[simple_type_element_name][:max_inclusive] = max_inclusive
+    hpxml_data_types_dict[simple_type_element_name][:min_exclusive] = min_exclusive
+    hpxml_data_types_dict[simple_type_element_name][:max_exclusive] = max_exclusive
+  end
+
+  # construct HPXMLvalidator.xml
+  hpxml_validator = XMLHelper.create_doc(version = '1.0', encoding = 'UTF-8')
+  root = XMLHelper.add_element(hpxml_validator, 'sch:schema')
+  XMLHelper.add_attribute(root, 'xmlns:sch', 'http://purl.oclc.org/dsdl/schematron')
+  XMLHelper.add_element(root, 'sch:title', 'HPXML Schematron Validator: HPXML.xsd')
+  name_space = XMLHelper.add_element(root, 'sch:ns')
+  XMLHelper.add_attribute(name_space, 'uri', 'http://hpxmlonline.com/2019/10')
+  XMLHelper.add_attribute(name_space, 'prefix', 'h')
+  pattern = XMLHelper.add_element(root, 'sch:pattern')
+
+  # construct complexType and group elements dictionary
+  complex_type_or_group_dict = {}
+  ['//xs:complexType', '//xs:group', '//xs:element'].each do |param|
+    base_elements_xsd_doc.xpath(param).each do |param_type|
+      next if param_type.name == 'element' && (not ['XMLTransactionHeaderInformation', 'ProjectStatus', 'SoftwareInfo'].include?(param_type.get('name')))
```

Comment:
I wish we didn't have to hardcode paths here, but I'm going to hold up the PR for this.

Reply:
I agree. If they weren't `xs:element` elements in `BaseElements.xsd`, they wouldn't have had to be hardcoded.

Reply:
Got it. Good enough for now. Marking resolved.